core/dbt/context/macro_resolver.py

```diff
@@ -0,0 +1,143 @@
+from typing import (
+    Dict, MutableMapping, Optional
+)
+from dbt.contracts.graph.parsed import ParsedMacro
+from dbt.exceptions import raise_duplicate_macro_name, raise_compiler_error
+from dbt.include.global_project import PROJECT_NAME as GLOBAL_PROJECT_NAME
+from dbt.clients.jinja import MacroGenerator
+
+MacroNamespace = Dict[str, ParsedMacro]
+
+
+# This class builds the MacroResolver by adding macros
+# to various categories for finding macros in the right order,
+# so that higher precedence macros are found first.
+class MacroResolver:
```

Comment:
is this net-new? Trying to understand how this worked before - can you shed a little bit of light @gshank ?

Reply:
Before the only place we actually resolved macro names was in the MacroNamespace object, which required that all macros be added to the object. I looked at refactoring that, but it wasn't feasible the way that it worked. So I split it up into a the MacroResolver (which will resolve macro names without all that overhead) and a TestMacroNamespace which uses the MacroResolver.  The goal was to eventually replace the current MacroNamespace entirely, but that work hasn't been done yet. Those are still high-value performance targets.  This is all because of the heavy overhead of adding every single macro in a project to every single test rendering context.